pkg/operation/common/extensions.go

```diff
@@ -292,6 +299,137 @@ func WaitUntilExtensionCRDeleted(
 	return nil
 }
 
+// RestoreExtensionWithDeployFunction deploys the extesnions resource with the passed in deployFunc and sets its operation annotation to wait-for-state.
+// It then restores the state of the extension resource from the ShootState, creates any required state resources and sets the operation annotation to restore.
+func RestoreExtensionWithDeployFunction(
+	ctx context.Context,
+	shootState *gardencorev1alpha1.ShootState,
+	c client.Client,
+	resourceKind string,
+	namespace string,
+	deployFunc func(ctx context.Context, operationAnnotation string) (extensionsv1alpha1.Object, error),
+) error {
+	extensionObj, err := deployFunc(ctx, v1beta1constants.GardenerOperationWaitForState)
+	if err != nil {
+		return err
+	}
+
+	if err := RestoreExtensionObjectState(ctx, c, shootState, namespace, extensionObj, resourceKind); err != nil {
+		return err
+	}
+
+	return AnnotateExtensionObjectWithOperation(ctx, c, extensionObj, v1beta1constants.GardenerOperationRestore)
+}
+
+// MigrateExtensionCR adds the migrate operation annotation to the extension CR.
+func MigrateExtensionCR(
+	ctx context.Context,
+	c client.Client,
+	newObjFunc func() extensionsv1alpha1.Object,
+	namespace string,
+	name string,
+) error {
+	obj := newObjFunc()
+	obj.SetNamespace(namespace)
+	obj.SetName(name)
+
+	if err := c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: name}, obj); err != nil {
+		if client.IgnoreNotFound(err) == nil {
+			return nil
+		}
+		return err
+	}
+
+	return AnnotateExtensionObjectWithOperation(ctx, c, obj, v1beta1constants.GardenerOperationMigrate)
+}
+
+// WaitUntilExtensionCRMigrated waits until the migrate operation for the extension resource is successful.
+func WaitUntilExtensionCRMigrated(
```

Comment:
Now that the function is not attached to the botanist any more - can we have a unit test for it?